// racer/lib/modes/lww.js
// Generated by CoffeeScript 1.3.1
var Lww, transaction;

transaction = require('../transaction.server');

module.exports = function(_arg) {
  var store;
  store = _arg.store;
  return new Lww(store);
};

Lww = function(store) {
  this._store = store;
  this._nextVer = 1;
};

Lww.prototype.startId = function(cb) {
  var startId;
  startId = this._startId || (this._startId = (+(new Date)).toString(36));
  return cb(null, startId);
};

Lww.prototype.commit = function(txn, cb) {
  var ver;
  ver = this._nextVer++;
  transaction.setVer(txn, ver);
  return this._store._finishCommit(txn, ver, cb);
};

Lww.prototype.flush = function(cb) {
  return cb(null);
};

Lww.prototype.version = function(cb) {
  return cb(null, this._nextVer - 1);
};

Lww.prototype.snapshotSince = function(_arg, cb) {
  var clientId, subs, ver;
  ver = _arg.ver, clientId = _arg.clientId, subs = _arg.subs;
  return this._store.fetch(clientId, subs, function(err, data) {
    if (err) {
      return cb(err);
    }
    return cb(null, {
      data: data
    });
  });
};

Lww.prototype.checkStartMarker = function(clientStartId, cb) {
  var err;
  if (clientStartId !== this._startId) {
    err = "clientStartId != startId (" + clientStartId + " != " + this._startId + ")";
    return cb(err);
  }
  return cb(null);
};
