// racer/examples/todos/index.js
var express, fs, gzip, racer, shared;
express = require('express');
gzip = require('connect-gzip');
fs = require('fs');
racer = require('racer');
shared = require('./shared');
module.exports = function(store) {
  var app, defaultGroup;
  exports.app = app = express.createServer().use(express.favicon()).use('/todos', gzip.staticGzip(__dirname));
  if (!store) {
    store = racer.createStore({
      redis: {
        db: 2
      },
      listen: app
    });
  }
  store.flush();
  racer.js({
    require: __dirname + '/shared',
    entry: __dirname + '/client.js'
  }, function(js) {
    return fs.writeFileSync(__dirname + '/script.js', js);
  });
  defaultGroup = {
    todos: {
      0: {
        id: 0,
        completed: true,
        text: 'This one is done already'
      },
      1: {
        id: 1,
        completed: false,
        text: 'Example todo'
      },
      2: {
        id: 2,
        completed: false,
        text: 'Another example'
      }
    },
    todoIds: [1, 2, 0],
    nextId: 3
  };
  app.get('/todos', function(req, res) {
    return res.redirect('/todos/racer');
  });
  app.get('/todos/:group', function(req, res) {
    var group;
    group = req.params.group;
    return store.subscribe({
      _group: "groups." + group + ".**"
    }, function(err, model) {
      model.setNull("groups." + group, defaultGroup);
      model.set('_group.todoList', model.arrayRef('_group.todos', '_group.todoIds'));
      return model.bundle(function(bundle) {
        var listHtml, todo;
        listHtml = ((function() {
          var _i, _len, _ref, _results;
          _ref = model.get('_group.todoList');
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            todo = _ref[_i];
            _results.push(shared.todoHtml(todo));
          }
          return _results;
        })()).join('');
        return res.send("<!DOCTYPE html>\n<title>Todos</title>\n<link rel=stylesheet href=style.css>\n<body>\n<div id=overlay></div>\n<!-- calling via timeout keeps the page from redirecting if an error is thrown -->\n<form id=head onsubmit=\"setTimeout(todos.addTodo, 0);return false\">\n  <h1>Todos</h1>\n  <div id=add><div id=add-input><input id=new-todo></div><input id=add-button type=submit value=Add></div>\n</form>\n<div id=dragbox></div>\n<div id=content><ul id=todos>" + listHtml + "</ul></div>\n<script>init=" + bundle + "</script>\n<script src=https://ajax.googleapis.com/ajax/libs/jquery/1.6.2/jquery.min.js></script>\n<script src=https://ajax.googleapis.com/ajax/libs/jqueryui/1.8.15/jquery-ui.min.js></script>\n<script src=script.js></script>");
      });
    });
  });
  return exports;
};
